operators/pkg/instance-controller/cloudinit.go

```diff
@@ -34,37 +36,48 @@ const (
 	WebdavSecretUsernameKey = "username"
 	// WebdavSecretPasswordKey -> The key of the webdav secret containing the password.
 	WebdavSecretPasswordKey = "password"
+
+	// UserDataKey -> the key of the created secret containing the cloud-init userdata content.
+	UserDataKey = "userdata"
 )
 
 // EnforceCloudInitSecret enforces the creation/update of a secret containing the cloud-init configuration,
 // based on the information retrieved for the tenant object and its associated WebDav credentials.
 func (r *InstanceReconciler) EnforceCloudInitSecret(ctx context.Context) error {
 	log := ctrl.LoggerFrom(ctx)
-	instance := clctx.InstanceFrom(ctx)
 
 	// Retrieve the WebDav credentials.
-	namespacedName := forge.NamespacedName(instance)
-	secretName := types.NamespacedName{Namespace: namespacedName.Namespace, Name: r.WebdavSecretName}
-	user, password, err := r.getWebDavCredentials(ctx, secretName)
+	user, password, err := r.GetWebDavCredentials(ctx)
 	if err != nil {
-		log.Error(err, "unable to get webdav credentials", "secret", secretName)
+		log.Error(err, "unable to get webdav credentials")
 		return err
 	}
-	log.V(utils.LogDebugLevel).Info("webdav credentials correctly retrieved", "secret", secretName)
+	log.V(utils.LogDebugLevel).Info("webdav credentials correctly retrieved")
 
 	// Retrieve the public keys
-	var publicKeys []string
-	if err = instance_creation.GetPublicKeys(ctx, r.Client, instance.Spec.Tenant, instance.Spec.Template, &publicKeys); err != nil {
+	publicKeys, err := r.GetPublicKeys(ctx)
+	if err != nil {
 		log.Error(err, "unable to get public keys")
 		return err
 	}
 	log.V(utils.LogDebugLevel).Info("public keys correctly retrieved")
 
-	// Create the cloud-init secret
-	secret := instance_creation.CreateCloudInitSecret(namespacedName.Name, namespacedName.Namespace, user, password, r.ServiceUrls.NextcloudBaseURL, publicKeys)
+	userdata, err := forge.CloudInitUserData(r.ServiceUrls.NextcloudBaseURL, user, password, publicKeys)
+	if err != nil {
+		log.Error(err, "unable to marshal secret content")
+		return err
+	}
+
+	// Enforce the cloud-init secret presence
+	instance := clctx.InstanceFrom(ctx)
+	secret := corev1.Secret{ObjectMeta: forge.ObjectMeta(instance)}
 	res, err := ctrl.CreateOrUpdate(ctx, r.Client, &secret, func() error {
+		secret.SetLabels(forge.InstanceObjectLabels(secret.GetLabels(), instance))
+		secret.Data = map[string][]byte{UserDataKey: userdata}
+		secret.Type = corev1.SecretTypeOpaque
```

Comment:
Is there no need to check for creation/update through the `.CreationTimestamp.IsZero()` thing?

Reply:
I personally don't see any risk in updating this, but I'm open to changes.

Reply:
No, although it is possible to mark a secret/configmap as immutable. I would personally enforce the latest version, so that in case of updates cloud-init can be re-run with the updated version (some hooks are run at every boot IIRC, and the others can be rerun manually if necessary). This would be useful for ssh keys for instance.